framework/include/utils/CellCenteredMapFunctor.h

```diff
@@ -0,0 +1,125 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseFunctor.h"
+#include "GreenGaussGradient.h"
+#include "MathFVUtils.h"
+#include "libmesh/utility.h"
+#include "libmesh/type_tensor.h"
+#include "libmesh/compare_types.h"
+#include "libmesh/threads.h"
+
+template <typename T, typename T2, typename std::enable_if<ScalarTraits<T>::value, int>::type = 0>
+inline TypeVector<typename CompareTypes<T, T2>::supertype>
+outer_product(const T & a, const TypeVector<T2> & b)
+{
+  TypeVector<typename CompareTypes<T, T2>::supertype> ret;
+  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+    ret(i) = a * b(i);
+
+  return ret;
+}
+
+template <typename T, typename Map>
+class CellCenteredMapFunctor : public Moose::FunctorImpl<T>, public Map
+{
+public:
+  using typename Moose::FunctorImpl<T>::ValueType;
+  using typename Moose::FunctorImpl<T>::GradientType;
+  using typename Moose::FunctorImpl<T>::DotType;
+
+  CellCenteredMapFunctor(const MooseMesh & mesh, const bool nonorthgonal_correction)
+    : _mesh(mesh), _nonorthgonal_correction(nonorthgonal_correction)
+  {
+  }
+
+  bool isExtrapolatedBoundaryFace(const FaceInfo & fi) const override { return !fi.neighborPtr(); }
+
+private:
+  const MooseMesh & _mesh;
+  const bool _nonorthgonal_correction;
+
+  ValueType evaluate(const Moose::ElemArg & elem_arg, unsigned int) const override final
+  {
+    const Elem * const elem = elem_arg.elem;
+    return libmesh_map_find(*this, elem->id());
+  }
+
+  ValueType evaluate(const Moose::FaceArg & face, unsigned int) const override final
+  {
+    const auto & fi = *face.fi;
+    mooseAssert(face.limiter_type == Moose::FV::LimiterType::CentralDifference,
+                "this implementation currently only supports linear interpolations");
+    if (fi.neighborPtr())
+      return Moose::FV::linearInterpolation(*this, face);
+    else
+    {
+      const auto elem_arg = face.makeElem();
+      const auto elem_value = (*this)(elem_arg);
+      // Two term expansion
+      return elem_value + this->gradient(elem_arg) * (fi.faceCentroid() - fi.elemCentroid());
+    }
+  }
+
+  using Moose::FunctorImpl<T>::evaluateGradient;
+
+  GradientType evaluateGradient(const Moose::ElemArg & elem_arg, unsigned int) const override final
+  {
+    return Moose::FV::greenGaussGradient(elem_arg, *this, true, _mesh);
+  }
+
+  GradientType evaluateGradient(const Moose::FaceArg & face, unsigned int) const override final
+  {
+    const auto & fi = *face.fi;
+    const auto elem_arg = face.makeElem();
+    const auto elem_gradient = this->gradient(elem_arg);
+    if (fi.neighborPtr())
+    {
+      const auto neighbor_arg = face.makeNeighbor();
+      const auto linear_interp_gradient =
+          fi.gC() * elem_gradient + (1 - fi.gC()) * this->gradient(neighbor_arg);
+      if (_nonorthgonal_correction)
```

Comment:
Hmm we have the same logic in `MooseVariableFV`. This should definitely be corrected, but I wonder whether that should be done in a follow-on as I think it could require some re-golding

Reply:
Actually I'll change this here. And I think I should rename `uncorrectedAdGradSln` to `arithmeticMeanFaceGrad` or something

Reply:
If you haven't changed already I think you should skip. I'm planning to add other non-orthogonal correction methods early next year and I could do this there at that point.

Reply:
In this code, I actually just removed the option to do "non-corrected".